File: 0451. Sort Characters By Frequency/solution.py
class Solution:
    # slow
    def frequencySort(self, s: str) -> str:
        charDic = {}
        for c in s:
            if c not in charDic:
                charDic[c] = 1
            else:
                charDic[c] += 1

        sortedCharDic = sorted(charDic.items(), key=lambda x: x[1], reverse=True)

        res = []
        for k, v in sortedCharDic:
            res += [k] * v

        return "".join(res)

    # online solution
    def frequencySort2(self, s: str) -> str:
        ans_str = ''
        # Find unique characters
        characters = set(s)

        counts = []
        # Count their frequency
        for i in characters:
            counts.append([i, s.count(i)])

        # Sort characters according to their frequency
        counts = sorted(counts, key=lambda x: x[1], reverse=True)

        # Generate answer string by multiplying frequency count with the character
        for i, j in counts:
            ans_str += i * j

        return ans_str

File: 0451. Sort Characters By Frequency/solution.java
class Solution {
    public static void main(String[] args) {
        System.
    }

    public String frequencySort(String s) {
        int[] charDic = new int[52];

        // store the frequency of the characters
        for(int i  = 0; i < s.length(); i++){
            int c = s.charAt(i);
            charDic[c - 65] += 1;
//            System.out.println(c);
        }

        String output = "";

        // "sort"
        while (true) {
            int maxFrequency = 0;
            char c = 0;
            for (int i = 0; i < charDic.length; i++) {
                int f = charDic[i];
                if (f > maxFrequency){
                    maxFrequency = f;
                    c = (char) (i + 65);
                }
            }

            if (maxFrequency == 0){
                break;
            }

            while(maxFrequency > 0){
                output += c;
                maxFrequency -= 1;
            }


        }

        return output;
    }
}